src/storage/engine/perf_context.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use rocksdb::PerfContext;
+
+// TODO: Can be simplified by const generics once RFC #2000 is landed.
+
+#[derive(Clone, Debug, Copy)]
+pub struct TypeDelta;
+#[derive(Clone, Debug, Copy)]
+pub struct TypeInstant;
+
+pub trait Type: fmt::Debug {}
+impl Type for TypeDelta {}
+impl Type for TypeInstant {}
+
+/// Store statistics we need. Data comes from RocksDB's `PerfContext`.
+///
+/// Depends on different contexts, it may either store instant values (i.e. absolute values
+/// at a specific time) or delta values (i.e. changed values during some time).
+#[derive(Clone, Debug, Copy)]
+pub struct PerfStatistics<T: Type> {
+    _phantom: PhantomData<T>,
+    pub internal_key_skipped_count: usize,
+    pub internal_delete_skipped_count: usize,
+    pub block_cache_hit_count: usize,
+    pub block_read_count: usize,
+    pub block_read_byte: usize,
+}
+
+/// Indicate that this statistics store delta values between two instant statistics.
+pub type PerfStatisticsDelta = PerfStatistics<TypeDelta>;
+
+/// Indicate that this statistics store instant values.
+pub type PerfStatisticsInstant = PerfStatistics<TypeInstant>;
+
+impl PerfStatisticsInstant {
+    /// Create an instance which stores instant statistics values, retrieved at creation.
+    pub fn new() -> Self {
+        let perf_context = PerfContext::get();
```

Comment:
Looks like it's in the global scope, is it possible that there may be some noise from other workflow(apply, split) if we collect it in the coprocessor?

Reply:
Don't worry. RocksDb's perf context is thread local.

Reply:
Then I think it should be `!Send`.

Reply:
@overvenus You are right!

Reply:
@overvenus Right. But adding `!Send` requires many refinement to the RequestTracker now, because `RequestTracker` is indeed being sent, only that the perf_context inside is not sent. I will add an issue to let contributors do the refinement (as well as adding !Send bound). Do you accept this solution?

Reply:
How about sending `PerfStatisticsDelta` instead of `PerfStatisticsInstant`? `impl !Send for TypeInstant`?

Reply:
We have to keep an instance of `Instant` in request tracker, so that later it can be used to calculate a delta, if we want it to be managed by request tracker.